### PR TITLE
Fix SHACL generator handling of special ifabsent functions

### LINKED ISSUE #3088: ifabsent processing is not implemented correctly in SHACL
---
name: Bug report
about: ''
title: 'linkml-model upgrade fails with linkml 1.9.6 due to ifabsent processing issue (SHACL)'
labels: bug
assignees: 'matentzn'

---

**Describe the bug**

SHACL generation _in linkml-model_ fails with linkml 1.9.6 due to ifabsent processing issue.

When running `gen-project` on the linkml-model schema, the SHACL generator fails with the following error:

```
Traceback (most recent call last):
  File ".venv/bin/gen-project", line 10, in <module>
    sys.exit(cli())
             ^^^^^
  File ".venv/lib/python3.11/site-packages/click/core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/click/core.py", line 1082, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/click/core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/click/core.py", line 788, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/linkml/generators/projectgen.py", line 256, in cli
    gen.generate(yamlfile, project_config)
  File ".venv/lib/python3.11/site-packages/linkml/generators/projectgen.py", line 146, in generate
    gen_dump = gen.serialize(**serialize_args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/linkml/generators/shaclgen.py", line 54, in serialize
    g = self.as_graph()
        ^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/linkml/generators/shaclgen.py", line 223, in as_graph
    default_value = ifabsent_processor.process_slot(s, c)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/linkml/generators/common/ifabsent_processor.py", line 87, in process_slot
    return self._map_to_default_value(slot, ifabsent_default_value, cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/linkml/generators/common/ifabsent_processor.py", line 211, in _map_to_default_value
    raise ValueError(f"The ifabsent value `{slot.ifabsent}` of the `{slot.name}` slot could not be processed")
ValueError: The ifabsent value `default_range` of the `range` slot could not be processed
```

## Root Cause

The `range` slot in the linkml-model meta schema uses `ifabsent: default_range`, which the SHACL generator's `ifabsent_processor` does not know how to handle.

## Workaround.. 

We can exclude the SHACL generator in `gen_project_config.yaml`, but this cyclic dependency is problemantic. Now even if we fix this issue in linkml repo, we need then create a new linkml release, then immediately another linkml-model release, then immediately another linkml-(runtime) release. 

## Expected Behavior

The SHACL generator should either:
1. Support the `default_range` ifabsent value, or
2. Gracefully skip/ignore slots with unsupported ifabsent values rather than raising an exception

**Version of LinkML you are using**

- linkml: 1.9.6
- linkml-runtime: 1.9.5
- Python: 3.11

**Please provide a schema (and if applicable, a data file) that replicates the issue**

NA

### PR DESCRIPTION
**Pull request by matentzn**
Closes #3088

Skip runtime-computed ifabsent values (`default_range`, `default_prefix`, `slot_uri`, `bnode`, etc.) instead of raising `ValueError`.

## The Problem

The SHACL generator failed when processing schemas with special ifabsent functions:

```
ValueError: The ifabsent value `default_range` of the `range` slot could not be processed
```

## Concern: Is this fix a bandaid?

A legitimate concern: Does skipping these values address the root cause, or should we resolve them to static values?

For example, `default_range` could theoretically be resolved by looking up `schema_view.schema.default_range` and using that as the SHACL default.

## Why skipping is the correct solution

The key insight is understanding what SHACL is for versus what ifabsent functions describe:

1. **SHACL validates instance data** against a schema. It answers: "Does this data conform to the schema?"

2. **ifabsent functions describe schema processing rules** - how the LinkML processor fills in missing values when loading schemas.

Consider `ifabsent: default_range` on the metamodel's `range` slot:
- This tells the LinkML processor: "If a schema doesn't specify a range, use the schema's `default_range`"
- This is a **schema processing rule**, not a **data validation rule**
- By the time SHACL validates data, all ranges are already resolved

Similarly:
- `bnode` means "generate a fresh blank node ID at runtime" - cannot be a static default
- `class_curie`, `slot_uri`, etc. describe how to derive URIs for schema elements, not data defaults

Skipping these (returning `None`) means no `sh:defaultValue` is emitted, which is semantically correct.

## The Fix

```python
RUNTIME_IFABSENT_FUNCTIONS = frozenset([
    "default_range",
    "default_prefix",
    "default_ns",
    "slot_curie",
    "slot_uri",
    "class_uri",
    "bnode",
])

def map_custom_default_values(self, default_value, slot, cls):
    if default_value == "class_curie":
        # ... existing handling ...

    if default_value in self.RUNTIME_IFABSENT_FUNCTIONS:
        return True, None

    return False, None
```

## Test Plan

- [x] `uv run pytest tests/linkml/test_generators/test_shaclgen.py` passes (19 tests)

### CODE REVIEW

File: packages/linkml/src/linkml/generators/shacl/shacl_ifabsent_processor.py [hunk 1]
```diff
@@ -6,12 +6,30 @@
...
+        ]
+    )
+
     def map_custom_default_values(self, default_value: str, slot: SlotDefinition, cls: ClassDefinition) -> (bool, str):
         if default_value == "class_curie":
             class_uri = self.schema_view.get_uri(cls, expand=True)
             if class_uri:
                 return True, URIRef(class_uri)
-            return True, ""
+            return True, None
```
Comment (cmungall) [inline packages/linkml/src/linkml/generators/shacl/shacl_ifabsent_processor.py:32]: type signature needs to change for this

**Comment by matentzn** [comment]
@HendrikBorgelt @jsheunis  @elichad I am tagging you all as you all have filed more than one issue on SHACL, and especially the SHACL generator. My mental model on the generator is super weak - I tried to have things explained to me by Claude as much as I could, but I would really value your feedback here so I can move needle on my attempt to re-sync with the latest linkml-model (metamodel).

File: packages/linkml/src/linkml/generators/shacl/shacl_ifabsent_processor.py (hunk 1 above)
Comment (matentzn) [reply packages/linkml/src/linkml/generators/shacl/shacl_ifabsent_processor.py:32]: https://github.com/linkml/linkml/commit/2e7734677a5985da34978f1ba9126b4fe3164e88

File: packages/linkml/src/linkml/generators/shacl/shacl_ifabsent_processor.py [hunk 2]
```diff
@@ -6,12 +8,32 @@
...
 class ShaclIfAbsentProcessor(IfAbsentProcessor):
-    def map_custom_default_values(self, default_value: str, slot: SlotDefinition, cls: ClassDefinition) -> (bool, str):
+    # Special ifabsent functions that compute values at runtime based on schema context.
+    # These cannot be represented as static SHACL default values, so we skip them.
+    RUNTIME_IFABSENT_FUNCTIONS = frozenset(
+        [
+            "default_range",
+            "default_prefix",
+            "default_ns",
+            "slot_curie",
```
Comment (sneakers-the-rat) [inline packages/linkml/src/linkml/generators/shacl/shacl_ifabsent_processor.py:18]: I believe excluding all the curie/uri values is incorrect - observe how class curie and class uri are actually processed in the following few lines.

**Comment by sneakers-the-rat** [comment]
If the AI description of why its choosing to return None here is correct, then there should be no handling of `ifabsent` in shacl at all. Notice how it justifies excluding `class_uri` but not `class_curie` despite those being logically the same thing (after expanding the curie) - presumably because there is a test for class curie and not class uri: https://github.com/linkml/linkml/blob/adc2d7d37d1a6f65696d93202d3d82292969c869/tests/linkml/test_generators/shacl/test_shacl_ifabsent_processor.py#L250
Im not sure why we should trust a language model's assessment of what should be done here when it is making assertions that are logically inconsistent on their face in order to just make tests pass.

Even though `default_range` is only used in one place (in the metamodel's range), and is an edge case that is different in kind than the normal usage of `ifabsent` - defining a default type rather than a default literal value - it still has a perfectly valid literal value: `xsd:string`. 

Since `default_range` is only really used in the metamodel, and the wonderful nature of the self defining schema means that schemas are merely data to the metamodel, then when validating the schema-as-data, it makes perfect sense to have a default value for `range` be `xsd:string`.

It would be a shame to let "AI" misunderstandings drive development decisions that compromise core language design features.

---

edit: @vincentkelleher was the original author of the ifabsent processors (including the shacl ifabsent processor) so probably could review this

**Comment by matentzn** [comment]
> Im not sure why we should trust a language model's assessment of what should be done here when it is making assertions that are logically inconsistent on their face in order to just make tests pass.

@sneakers-the-rat you are absolutely right. My plan is to see if I can form some form of "expert panels" around the different generators, and produce some community policies around responsible use of AI for LinkML; we would not have merged this "AI slop" without some human expert review. 

I am primarily working on extending and improving the test system in LinkML - I have a very limited idea of how the generators work. This PR was special as it severely blocks a whole cascade of things I need to do (LinkML-Model upgrades, integration tests etc) - so I gave it a shot. I am glad I missed and you shot back to stop me :) So thanks! I will look to see if I can get your proposal in the issue to work instead.
